workflow/controller/controller.go

```diff
@@ -534,6 +534,65 @@ func (wfc *WorkflowController) signalContainers(namespace string, podName string
 	return time.Duration(*pod.Spec.TerminationGracePeriodSeconds) * time.Second, nil
 }
 
+func (wfc *WorkflowController) deleteOffloadedNodes() error {
+	oldRecords, err := wfc.offloadNodeStatusRepo.ListOldOffloads(wfc.GetManagedNamespace())
+	if err != nil {
+		return err
+	}
+	if len(oldRecords) == 0 {
+		log.Info("Zero old offloads, nothing to do")
+		return nil
+	}
+	workflows, err := util.NewWorkflowLister(wfc.wfInformer).List()
+	if err != nil {
+		return err
+	}
+	for _, wf := range workflows {
+		log.WithField("uid", wf.UID).Debug("Locking workflow")
+		key, err := cache.MetaNamespaceKeyFunc(wf)
+		if err != nil {
+			return err
+		}
+		wfc.workflowKeyLock.Lock(key)
+	}
+	defer func() {
+		for _, wf := range workflows {
+			key, err := cache.MetaNamespaceKeyFunc(wf)
+			if err == nil {
+				log.WithField("uid", wf.UID).Debug("Unlocking workflow")
+				wfc.workflowKeyLock.Unlock(key)
+			}
+		}
+	}()
+	// get every lives workflow (1000s) into a map
+	liveOffloadNodeStatusVersions := make(map[types.UID]string)
+	for _, wf := range workflows {
+		// workflow might still be hydrated
+		if wfc.hydrator.IsHydrated(wf) {
+			log.WithField("uid", wf.UID).Debug("Hydrated workflow encountered")
```

Comment:
rather than dehydrating, instead you should just skip the workflow, clean up will occur on next loop
I think it would be good to log at info level too

Reply:
I've observed that often the GC keeps receiving the same hydrated workflow from the lister. Even long after the workflow completed. Wouldn't skipping these workflows not result in garbage pilling up?

Reply:
ah - good point - after deletion it could remain hydrated - keep this, but lets log at info